### PR TITLE
[SYCL][DOC] Adding design document for command graph extension

### PR DESCRIPTION
**Pull request by reble**
This is an initial draft of our implementation design documentation.

### CODE REVIEW

File: sycl/doc/design/CommandGraph.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+#### Command Buffer extension
+
+| Function | Description |
+| ------------------------- | ------------------------ |
+| piextCommandBufferCreate: | create a command-buffer. |
+| piextCommandBufferRetain: | incrementing reference count of command-buffer. |
+| piextCommandBufferRelease: decrementing reference count of command-buffer. |
+| piextCommandBufferFinalize: no more commands can be appended, makes command
+ buffer ready to enqueue on command-queue. |
+| piextCommandBufferNDRangeKernel: append a kernel execution command to command
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]: Missing the middle `|` column separator

File: sycl/doc/design/CommandGraph.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+This document describes the implementation design of the 
+[SYCL Graph Extension](https://github.com/intel/llvm/pull/5626).
+
+A related presentation can be found 
+[here](https://www.youtube.com/watch?v=aOTAmyr04rM).
+
+## Requirements
+
+An efficient implementation of a lazy command graph execution and its replay
+requires extensions to the PI layer. Such an extension is command buffers.
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]: "Such an extension is command buffers.", we should expand on this sentence to say what a command-buffer is. e.g something like 

> Such an extension is command buffers, where a command-buffer object represents a series of operations to be enqueued to the backend device and their dependencies. A single command graph can be partitioned into more than one PI command-buffer by the runtime.

File: sycl/doc/design/CommandGraph.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+The emulation mode targets support of functionality only, without potentially
+resulting performance improvements, i.e. execution of a closed Level Zero
+command list multiple times. 
+
+#### Command Buffer extension
+
+| Function | Description |
+| ------------------------- | ------------------------ |
+| piextCommandBufferCreate: | create a command-buffer. |
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]: Would put backticks round the API names to render as monospace

**Review by EwanC — COMMENTED** [review]
Some other things I think it'd be worth documenting. Could be done in this PR, or in a follow-up: 

* Relationship with a handler for processing command-groups to use in a node - had to fix this for capturing of temporaries bug.
* Sub-graphs - we create an empty node as an implementation detail to marshal all the dependencies.
* breadth-first approach - Once #182 is merged we should write down that implementation detail.

File: sycl/doc/design/CommandGraph.md (hunk 2 above)
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
requires extensions to the PI layer. Such an extension is command buffers,
where a command-buffer object represents a series of operations to be enqueued
to the backend device and their dependencies. A single command graph can be
partitioned into more than one PI command-buffer by the runtime.
```

File: sycl/doc/design/CommandGraph.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+| piextCommandBufferRetain: | incrementing reference count of command-buffer. |
+| piextCommandBufferRelease: decrementing reference count of command-buffer. |
+| piextCommandBufferFinalize: no more commands can be appended, makes command
+ buffer ready to enqueue on command-queue. |
+| piextCommandBufferNDRangeKernel: append a kernel execution command to command
+buffer. |
+| piextEnqueueCommandBuffer: | submit command-buffer to queue for execution |
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
| `piextCommandBufferCreate` | create a command-buffer. |
| `piextCommandBufferRetain` | incrementing reference count of command-buffer. |
| `piextCommandBufferRelease` | decrementing reference count of command-buffer. |
| `piextCommandBufferFinalize` | no more commands can be appended, makes command
 buffer ready to enqueue on command-queue. |
| `piextCommandBufferNDRangeKernel` | append a kernel execution command to command
buffer. |
| `piextEnqueueCommandBuffer` | submit command-buffer to queue for execution |
```

File: sycl/doc/design/CommandGraph.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+#### Command Buffer extension
+
+| Function | Description |
+| ------------------------- | ------------------------ |
+| `piextCommandBufferCreate` | create a command-buffer. |
+| `piextCommandBufferRetain` | incrementing reference count of command-buffer. |
+| `piextCommandBufferRelease` | decrementing reference count of command-buffer. |
+| `piextCommandBufferFinalize` | no more commands can be appended, makes command buffer ready to enqueue on command-queue. |
+| `piextCommandBufferNDRangeKernel` | append a kernel execution command to command buffer. |
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md:33]: We also now have:
* piextCommandBufferMemcpyUSM - append a USM memcpy command to the command-buffer
* piextCommandBufferMemBufferCopy - append a mem buffer copy command to the command-buffer
* piextCommandBufferMemBufferCopyRect - append a rectangular mem buffer copy command to the command-buffer

File: sycl/doc/design/CommandGraph.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+## Requirements
+
+An efficient implementation of a lazy command graph execution and its replay
+requires extensions to the PI layer. Such an extension is command buffers,
+where a command-buffer object represents a series of operations to be enqueued
+to the backend device and their dependencies. A single command graph can be
+partitioned into more than one PI command-buffer by the runtime.
+We distinguish between backends that support command buffer extensions and
+those that do not. Currently command buffer extensions are only supported by
+Level Zero. All other backends would fall back to an emulation mode.
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]: "All other backends would fall back to an emulation mode, **or not be reported as supported**"

File: sycl/doc/design/CommandGraph.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+The emulation mode targets support of functionality only, without potentially
+resulting performance improvements, i.e. execution of a closed Level Zero
+command list multiple times. 
+
+#### Command Buffer extension
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
### Command Buffer extension
```

**Review by EwanC — APPROVED** [review]
This is a good starting point.

Would also like to have sections with my previous suggestions, and also taking the comment content from https://github.com/reble/llvm/blob/452e7aebc5af88ee38d5836a04513b67348ede83/sycl/plugins/level_zero/pi_level_zero.cpp#L8832-L8889 

But can write all that up in a follow-up PR.

File: sycl/doc/design/CommandGraph.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+| `piextEnqueueCommandBuffer` | submit command-buffer to queue for execution |
+
+## Design
+
+![Basic architecture diagram.](images/SYCL-Graph-Architecture.svg)
+
+There are two sets of user facing interfaces that are proposed to create a
+command graph object: 
+Explicit and Record & Replay API. Within the runtime they share a common
+infrastructure.
```
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
There are two sets of user facing interfaces that can be used to create a
command graph object: 
Explicit and Record & Replay API. Within the runtime they share a common
infrastructure.
```
Nitpick but "proposed" sounds more like we're pitching the extension specification, where this document is just talking about what we actually implement.

File: sycl/doc/design/CommandGraph.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+proposal. Memory operations will be supported subsequently by the current
+implementation starting with `memcpy`.
+
+Buffers and accessors are supported in a command graph. Following restrictions
+are required to adapt buffers and their lifetime to a lazy work execution model:
+
+- Lifetime of a buffer with host data will be extended by copying the underlying
+data.
+- Host accessor on buffer that are used by a command graph are prohibited.
+- Copy-back behavior on destruction of a data is prohibited.
```
Comment (Bensuo) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
- Copy-back behavior on destruction of a buffer is prohibited. 
```

**Review by Bensuo — APPROVED** [review]

File: sycl/doc/design/CommandGraph.md (hunk 6 above)
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
Level Zero. All other backends would fall back to an emulation mode, or not
be reported as supported.
```

File: sycl/doc/design/CommandGraph.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+A related presentation can be found 
+[here](https://www.youtube.com/watch?v=aOTAmyr04rM).
+
+## Requirements
+
+An efficient implementation of a lazy command graph execution and its replay
+requires extensions to the PI layer. Such an extension is command buffers,
+where a command-buffer object represents a series of operations to be enqueued
+to the backend device and their dependencies. A single command graph can be
+partitioned into more than one PI command-buffer by the runtime.
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
partitioned into more than one command-buffer by the runtime.
```

File: sycl/doc/design/CommandGraph.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+This document describes the implementation design of the 
+[SYCL Graph Extension](https://github.com/intel/llvm/pull/5626).
+
+A related presentation can be found 
+[here](https://www.youtube.com/watch?v=aOTAmyr04rM).
+
+## Requirements
+
+An efficient implementation of a lazy command graph execution and its replay
+requires extensions to the PI layer. Such an extension is command buffers,
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md]:
```suggestion
requires extensions to the UR layer. Such an extension is command buffers,
```

File: sycl/doc/design/CommandGraph.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+| Function | Description |
+| ------------------------- | ------------------------ |
+| `piextCommandBufferCreate` | create a command-buffer. |
+| `piextCommandBufferRetain` | incrementing reference count of command-buffer. |
+| `piextCommandBufferRelease` | decrementing reference count of command-buffer. |
+| `piextCommandBufferFinalize` | no more commands can be appended, makes command buffer ready to enqueue on command-queue. |
+| `piextCommandBufferNDRangeKernel` | append a kernel execution command to command buffer. |
+| `piextEnqueueCommandBuffer` | submit command-buffer to queue for execution |
+
```
Comment (reble) [inline sycl/doc/design/CommandGraph.md:38]:
```suggestion
| `piextCommandBufferMemcpyUSM` | append a USM memcpy command to the command-buffer. |
| `piextCommandBufferMemBufferCopy` | append a mem buffer copy command to the command-buffer. |
| `piextCommandBufferMemBufferCopyRect` | append a rectangular mem buffer copy command to the command-buffer. |

```

File: sycl/doc/design/CommandGraph.md (hunk 5 above)
Comment (reble) [reply sycl/doc/design/CommandGraph.md:33]: done

File: sycl/doc/design/CommandGraph.md (hunk 6 above)
Comment (reble) [reply sycl/doc/design/CommandGraph.md]: done

**Pull request merged**